muxer.go

```diff
@@ -0,0 +1,411 @@
+	w          io.Writer
+	bitsWriter *astikit.BitsWriter
+
+	packetSize             int
+	tablesRetransmitPeriod int // period in PES packets
+
+	pm         programMap // pid -> programNumber
+	pmt        PMTData
+	nextPID    uint16
+	patVersion wrappingCounter
+	pmtVersion wrappingCounter
+
+	patBytes bytes.Buffer
+	pmtBytes bytes.Buffer
+
+	buf       bytes.Buffer
+	bufWriter *astikit.BitsWriter
+
+	esContexts              map[uint16]*esContext
+	tablesRetransmitCounter int
+}
+
+type esContext struct {
+	es *PMTElementaryStream
+	cc wrappingCounter
+}
+
+func newEsContext(es *PMTElementaryStream) *esContext {
+	return &esContext{
+		es: es,
+		cc: newWrappingCounter(0b1111), // CC is 4 bits
+	}
+}
+
+func MuxerOptionTablesRetransmitPeriod(newPeriod int) func(*Muxer) {
+	return func(m *Muxer) {
+		m.tablesRetransmitPeriod = newPeriod
+	}
+}
+
+func NewMuxer(ctx context.Context, w io.Writer, opts ...func(*Muxer)) *Muxer {
+	m := &Muxer{
+		ctx: ctx,
+		w:   w,
+
+		packetSize:             MpegTsPacketSize, // no 192-byte packet support yet
+		tablesRetransmitPeriod: 40,
+
+		pm: newProgramMap(),
+		pmt: PMTData{
+			ElementaryStreams: []*PMTElementaryStream{},
+			ProgramNumber:     ProgramNumberStart,
+		},
+
+		// table version is 5-bit field
+		patVersion: newWrappingCounter(0b11111),
+		pmtVersion: newWrappingCounter(0b11111),
+
+		esContexts: map[uint16]*esContext{},
+	}
+
+	m.bufWriter = astikit.NewBitsWriter(astikit.BitsWriterOptions{Writer: &m.buf})
+	m.bitsWriter = astikit.NewBitsWriter(astikit.BitsWriterOptions{Writer: m.w})
+
+	// TODO multiple programs support
+	m.pm.set(PMTStartPID, ProgramNumberStart)
+
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	// to output tables at the very start
+	m.tablesRetransmitCounter = m.tablesRetransmitPeriod
+
+	return m
+}
+
+// if es.ElementaryPID is zero, it will be generated automatically
+func (m *Muxer) AddElementaryStream(es PMTElementaryStream, isPCRPid bool) error {
+	if es.ElementaryPID != 0 {
+		for _, oes := range m.pmt.ElementaryStreams {
+			if oes.ElementaryPID == es.ElementaryPID {
+				return MuxerErrorPIDAlreadyExists
+			}
+		}
+	} else {
+		es.ElementaryPID = m.nextPID
+		m.nextPID++
+	}
+
+	m.pmt.ElementaryStreams = append(m.pmt.ElementaryStreams, &es)
+	if isPCRPid {
+		m.pmt.PCRPID = es.ElementaryPID
+	}
+
+	m.esContexts[es.ElementaryPID] = newEsContext(&es)
+	// invalidate pmt cache
+	m.pmtBytes.Reset()
+	return nil
+}
+
+func (m *Muxer) RemoveElementaryStream(pid uint16) error {
+	foundIdx := -1
+	for i, oes := range m.pmt.ElementaryStreams {
+		if oes.ElementaryPID == pid {
+			foundIdx = i
+			break
+		}
+	}
+
+	if foundIdx == -1 {
+		return MuxerErrorPIDNotFound
+	}
+
+	m.pmt.ElementaryStreams = append(m.pmt.ElementaryStreams[:foundIdx], m.pmt.ElementaryStreams[foundIdx+1:]...)
+	delete(m.esContexts, pid)
+	m.pmtBytes.Reset()
+	return nil
+}
+
+func (m *Muxer) WritePayload(pid uint16, af *PacketAdaptationField, ph *PESHeader, payload []byte) (int, error) {
```

Comment:
I think the public API should be `func (m *Muxer) WriteData(af *PacketAdaptationField, d *Data) (int, error)` instead. I don't mind implementing it for PES only for now. `*Data` would be sufficient to pass `pid`, `ph` and `payload`

Reply:
I gotta remember why I haven't done that in the first place, will try to remember in next couple of days and will get back to you :)

Reply:
Aha! I remember now why I haven't done it that way.

- `Data` has `FirstPacket` with `PID` in its header. At the same time `Data` has `PID` as a separate member. While it's perfectly okay and handy for demuxer API user, that is confusing for muxer API user
- `FirstPacket` has a bunch of fields which have no effect setting in `WriteData` case - I will rewrite CC, AF/payload indicators, etc. That could confuse API user as well.

So the main point here is: I've designed `WritePayload` to require arguments that really make sense here and are really used.

I get your point as well - it's not consistent with demuxer currently. And it's not especially pretty :) Though I don't really know a way to make it perfect now - seems like we're choosing between two tradeoffs (well, a position I constantly find myself in :))

Reply:
Also I think muxer logic is a bit different by its nature.

Demuxer usage is like "tell me everything you can parsing a TS stream".

While muxer usage is more like "okay, I got a bunch of elementary streams, now let's write payloads. And handle all the low level stuff like PAT/PMT for me, I don't want to be bothered by that".

Reply:
Love your explanation and I get your point about `FirstPacket`, that would indeed be confusing.

What about renaming the current `Data` to `DemuxerData` and creating :

```go
type MuxerData struct {
  AdaptationField *PacketAdaptationField
  PES *PESData
  PID uint16
}
```

I think that would be a good compromise.

Reply:
Let me think about it a little bit.

Reply:
Well I've done it that way and I like it.

Though don't rush to resolve this thread just yet. I'd like to do some additional testing today to make sure I haven't broken anything.

Reply:
I've tested new `WriteData` with my project I'm working on. All is ok, no issues.

Reply:
Nice! I'll merge the PR then